pkg/ring/kv/dynamodb/client.go

```diff
@@ -0,0 +1,313 @@
+
+	"github.com/cortexproject/cortex/pkg/ring/kv/codec"
+	"github.com/cortexproject/cortex/pkg/util/backoff"
+)
+
+// Config to create a ConsulClient
+type Config struct {
+	Region    string        `yaml:"region"`
+	TableName string        `yaml:"table_name"`
+	TTL       time.Duration `yaml:"ttl"`
+}
+
+type Client struct {
+	kv         dynamoDbClient
+	codec      codec.Codec
+	ddbMetrics *dynamodbMetrics
+	logger     log.Logger
+
+	staleDataLock sync.RWMutex
+	staleData     map[string]staleData
+}
+
+type staleData struct {
+	data      codec.MultiKey
+	timestamp time.Time
+}
+
+var (
+	backoffConfig = backoff.Config{
+		MinBackoff: 1 * time.Second,
+		MaxBackoff: 1 * time.Minute,
+		MaxRetries: 0,
+	}
+
+	defaultLoopDelay = 1 * time.Minute
+)
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+// If prefix is not an empty string it should end with a period.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet, prefix string) {
+	f.StringVar(&cfg.Region, prefix+"dynamodb.region", "", "Region to access dynamodb.")
+	f.StringVar(&cfg.TableName, prefix+"dynamodb.table-name", "", "Table name to use on dynamodb.")
+	f.DurationVar(&cfg.TTL, prefix+"dynamodb.ttl-time", 0, "Time to expire items on dynamodb.")
+}
+
+func NewClient(cfg Config, cc codec.Codec, logger log.Logger, registerer prometheus.Registerer) (*Client, error) {
+	dynamoDB, err := newDynamodbKV(cfg, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	ddbMetrics := newDynamoDbMetrics(registerer)
+
+	c := &Client{
+		kv:         dynamodbInstrumentation{kv: dynamoDB, ddbMetrics: ddbMetrics},
+		codec:      cc,
+		logger:     logger,
+		ddbMetrics: ddbMetrics,
+		staleData:  make(map[string]staleData),
+	}
+	return c, nil
+}
+
+func (c *Client) List(ctx context.Context, key string) ([]string, error) {
+	resp, _, err := c.kv.List(ctx, dynamodbKey{primaryKey: key})
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "error listing dynamodb", "key", key, "err", err)
+		return nil, err
+	}
+	return resp, err
+}
+
+func (c *Client) Get(ctx context.Context, key string) (interface{}, error) {
+	resp, _, err := c.kv.Query(ctx, dynamodbKey{primaryKey: key}, false)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "error Get dynamodb", "key", key, "err", err)
+		return nil, err
+	}
+	res, err := c.decodeMultikey(resp)
+	if err != nil {
+		return nil, err
+	}
+	c.updateStaleData(key, res)
+
+	return res, nil
+}
+
+func (c *Client) Delete(ctx context.Context, key string) error {
+	resp, _, err := c.kv.Query(ctx, dynamodbKey{primaryKey: key}, false)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "error Delete dynamodb", "key", key, "err", err)
+		return err
+	}
+
+	for innerKey := range resp {
+		err := c.kv.Delete(ctx, dynamodbKey{
+			primaryKey: key,
+			sortKey:    innerKey,
+		})
+		if err != nil {
+			level.Warn(c.logger).Log("msg", "error Delete dynamodb", "key", key, "innerKey", innerKey, "err", err)
+			return err
+		}
+	}
+	return err
```

Comment:
err here is always nil, we can change it to return nil

Reply:
Do we need to deleteStaleData(key) here?

Reply:
We can change the to be nil. It should not be a difference.

Good point about the deleteStaleData. I will add it